DevEx #537 Replace VueX with Pinia

Haviles04 opened the pull request:
<!-- Thanks for contributing to Cuttle! 🎉 -->

## Issue number

Relevant [issue number](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- Resolves #537

## Please check the following

- [x] Do the tests still pass? (see [Run the Tests](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#run-the-tests))
- [x] Is the code formatted properly? (see [Linting (Formatting)](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#linting-formatting))
- For New Features:
  - [ ] Have tests been added to cover any new features or fixes?
  - [ ] Has the documentation been updated accordingly?

## Please describe additional details for testing this change\

Migrated entire application store from VueX to Pinia.
-  Changed store modules to flat file structure 
- Converted old modules to new functional pinia modules
- Removed Actions that only set to a new value, and set it explicitly where the action was used
- Changed store references throughout entire application to utilize new `...mapStores(useStore)` function
- Added store to window object on `mounted()` inside `App.vue`
- Changed entire test suite to utilize new window store `cy.window().its('storeName').then((store) => logic)`

itsalaidbacklife reviewed (commented):
Nice start. This is looking good so far. I left a couple comments, all minor. I think it may make sense to do one module and update its references ie get the references to the auth store module next

Haviles04 commented on src/stores/auth.js:
```diff
@@ -1,18 +1,19 @@
 };
 
-async function handleLogin(context, username, password, signup = false) {
+async function handleLogin(username, password, signup = false) {
```
Yea, I actually wasn't sure how this was going to work out. I meant to ask you. I like moving it to actions 🤙

Haviles04 commented on src/stores/auth.js:
```diff
@@ -1,80 +1,45 @@
+  }),
+  actions: {
+    authSuccess(username) {
+      this.state.authenticated = true;
+      this.state.username = username;
```
Need to remove state here

Haviles04 commented on src/stores/auth.js:
```diff
@@ -99,24 +64,22 @@ export default {
+              this.mustReauthenticate = false;
               const pNum =
-                res.pNum ?? getPlayerPnumByUsername(context.rootState.game.players, context.state.username);
+                res.pNum ?? getPlayerPnumByUsername(this.rootState.game.players, this.state.username);
```
Need to figure out rootState here

Haviles04 commented:
I made a bunch of progress on this, I changed all the vueX store modules to pinia, updated router component, updated socket event files, navLink composables, gameView, and loginView..

I'm unsure if I'm reading this correctly or not, but pinia guide says not to useStore outside of a function outside of a component, this is causing me to have to declare the store in every function inside `/plugins/sockets/connectivityEvents.js`
https://pinia.vuejs.org/core-concepts/outside-component-usage.html

itsalaidbacklife commented on src/App.vue:
```diff
@@ -31,7 +34,15 @@ export default {
+  },
+  created() {
+    //Pass store to window object on testing
+    if (window.cypress !== null) { 
```
I think it's a capital C and that it will be undefined rather than null when missing
```suggestion
    if (window.Cypress) { 
```

itsalaidbacklife commented on src/components/GameView/GameDialogs.vue:
```diff
@@ -99,57 +89,57 @@ export default {
       },
       set(val) {
-        this.$store.commit('setMustReauthenticate', val);
+        this.authStore.setMustReauthenticate(val);
```
It probably makes sense to remove any former mutation functions that just set a single variable and replace them with inline assignments
```suggestion
        this.authStore.mustReauthenticate = val;
```

itsalaidbacklife commented on src/components/GameView/FourDialog.vue:
```diff
@@ -64,7 +67,7 @@ export default {
     },
     hand() {
-      return this.$store.state.game.players[this.$store.state.game.myPNum].hand;
+      return this.gameStore.players[this.gameStore.myPNum].hand;
```
We can use the player getter to dry this up 
```suggestion
      return this.gameStore.player.hand;
```

itsalaidbacklife commented on src/components/GameView/GameDialogs.vue:
```diff
@@ -69,23 +69,13 @@ export default {
-      'opponentQueenCount',
-      'playerWins',
-    ]),
+    ...mapStores(useGameStore),
+    ...mapStores(useAuthStore),
```
mapStores can input multiple stores
```suggestion
    ...mapStores(useGameStore, useAuthStore),
```

itsalaidbacklife commented on src/components/GameView/GameDialogs.vue:
```diff
@@ -69,23 +69,13 @@ export default {
       // bit unusual-- the usages can probably be changed to only need a subset,
       // or moved to some sort of store method or getter
-      return this.$store.state.game;
+      return this.gameStore;
```
It feels a little weird to have this computed prop just to alias the gameStore that we already mapped. How about instead we replace the references to `game` with `gameStore`?

itsalaidbacklife commented on src/components/GameView/GameMenu.vue:
```diff
@@ -145,24 +148,24 @@ export default {
-        await this.$store.dispatch('requestStalemate');
-        this.$store.commit('setWaitingForOpponentToStalemate', true);
+        await this.gameStore.requestStalemate();
+        this.gameStore.setWaitingForOpponentToStalemate(true);
```
This looks like another method that just sets a single value; if so, let's remove it
```suggestion
        this.gameStore.waitingForOpponentToStalemate = true;
```

itsalaidbacklife reviewed (commented):
Awesome work. Commenting with my initial thoughts on the files I got through

Haviles04 commented on src/App.vue:
```diff
@@ -31,7 +34,15 @@ export default {
   }
- }
+  },
+  created() {
+    //Pass store to window object on testing
+    if (window.Cypress !== null) { 
+    window.gameStore = useGameStore();
+    window.authStore = useAuthStore();
+    window.gameListStore = useGameListStore();
+    }
```
I'm wondering if it would be better to do 
```
window.store = { 
 gameStore : useGameStore(),
 authStore : useAuthStore(),
 gameListStore: useGameListStore(),
 };

itsalaidbacklife replied:
Interesting idea. Ultimately it doesn't matter much either way, but I think it's more in the spirit of pinia to keep it flat like you already have it since that's closer to how you'd grab them with script setup. And really, what's the point in an additional layer of property access?

Haviles04 replied:
Cool, my thought was maybe putting it inside window.store would keep them all together, but you right it doesn't really matter. :)

seriouslysean replied:
It might still be better to keep these in the cuttle namespace regardless of the store nesting, right now you're polluting the window object.

itsalaidbacklife reviewed (commented):
Looking really good. Left a couple nitpick comments

itsalaidbacklife reviewed (approved):
Very nicely done. This will be a significant QoL improvement for DevEx

seriouslysean reviewed (approved):
Major QOL, nice work 👍
